File: apps/simulation/internal/model/pid.go
package model

import "time"

type PIDConfig struct {
	ControllerTag          string  `json:"controllerTag"`
	ProcessVariableTag     string  `json:"processVariableTag"`
	ManipulatedVariableTag string  `json:"manipulatedVariableTag"`
	Setpoint               float64 `json:"setpoint"`
	Kp                     float64 `json:"kp"`
	Ki                     float64 `json:"ki"`
	Kd                     float64 `json:"kd"`
	OutputMin              float64 `json:"outputMin"`
	OutputMax              float64 `json:"outputMax"`
	IntegralMin            float64 `json:"integralMin"`
	IntegralMax            float64 `json:"integralMax"`
	SampleTimeMS           int     `json:"sampleTimeMs"`
	Enabled                bool    `json:"enabled"`
}

type PIDState struct {
	Setpoint      float64   `json:"setpoint"`
	ProcessValue  float64   `json:"processValue"`
	Error         float64   `json:"error"`
	PreviousError float64   `json:"previousError"`
	Integral      float64   `json:"integral"`
	Derivative    float64   `json:"derivative"`
	PTerm         float64   `json:"pTerm"`
	ITerm         float64   `json:"iTerm"`
	DTerm         float64   `json:"dTerm"`
	Output        float64   `json:"output"`
	LastOutput    float64   `json:"lastOutput"`
	OutputBias    float64   `json:"outputBias"`
	LastUpdateAt  time.Time `json:"lastUpdateAt"`
	Active        bool      `json:"active"`
	Saturated     bool      `json:"saturated"`
	Status        string    `json:"status"`
}

type PIDStatus struct {
	ControllerTag          string           `json:"controllerTag"`
	Mode                   ControlMode      `json:"mode"`
	Authority              ControlAuthority `json:"authority"`
	Active                 bool             `json:"active"`
	PIDImplemented         bool             `json:"pidImplemented"`
	ProcessVariableTag     string           `json:"processVariableTag"`
	ProcessValue           float64          `json:"processValue"`
	Setpoint               float64          `json:"setpoint"`
	ManipulatedVariableTag string           `json:"manipulatedVariableTag"`
	Output                 float64          `json:"output"`
	OutputMin              float64          `json:"outputMin"`
	OutputMax              float64          `json:"outputMax"`
	Kp                     float64          `json:"kp"`
	Ki                     float64          `json:"ki"`
	Kd                     float64          `json:"kd"`
	Error                  float64          `json:"error"`
	PTerm                  float64          `json:"pTerm"`
	ITerm                  float64          `json:"iTerm"`
	DTerm                  float64          `json:"dTerm"`
	Integral               float64          `json:"integral"`
	Derivative             float64          `json:"derivative"`
	Saturated              bool             `json:"saturated"`
	Status                 string           `json:"status"`
	UpdatedAt              time.Time        `json:"updatedAt"`
	SafetyDisclaimer       string           `json:"safetyDisclaimer"`
}

type PIDConfigUpdateRequest struct {
	Setpoint    *float64 `json:"setpoint,omitempty"`
	Kp          *float64 `json:"kp,omitempty"`
	Ki          *float64 `json:"ki,omitempty"`
	Kd          *float64 `json:"kd,omitempty"`
	OutputMin   *float64 `json:"outputMin,omitempty"`
	OutputMax   *float64 `json:"outputMax,omitempty"`
	RequestedBy string   `json:"requestedBy,omitempty"`
	Reason      string   `json:"reason,omitempty"`
}

File: apps/web/src/test/fixtures/pid.ts
import type { PIDStatus } from "@/entities/pid/model/types";

export const pidManualInactiveFixture: PIDStatus = {
  controllerTag: "TIC-101",
  mode: "MANUAL",
  authority: "USER",
  active: false,
  pidImplemented: true,
  processVariableTag: "TT-101",
  processValue: 286.2,
  setpoint: 288,
  manipulatedVariableTag: "V-101.POS",
  output: 52.4,
  outputMin: 0,
  outputMax: 100,
  kp: 0.8,
  ki: 0.05,
  kd: 0.1,
  error: 1.8,
  pTerm: 1.44,
  iTerm: 0.4,
  dTerm: 0.05,
  integral: 8,
  derivative: 0.5,
  saturated: false,
  updatedAt: "2026-05-21T06:00:00Z",
  safetyDisclaimer:
    "TIC-101 is a simulation-only PID controller for a synthetic thermal loop. It does not control real equipment.",
  status: "Manual",
};

export const pidAutoActiveFixture: PIDStatus = {
  ...pidManualInactiveFixture,
  mode: "AUTO",
  authority: "PID",
  active: true,
  output: 61.7,
  status: "Active",
};

export const pidSaturatedFixture: PIDStatus = {
  ...pidAutoActiveFixture,
  output: 100,
  saturated: true,
  status: "Saturated",
};
